infrastructure/src/main/java/org/corfudb/infrastructure/management/ClusterStateAggregator.java

```diff
@@ -62,10 +66,41 @@ public ClusterState getAggregatedState() {
                         NodeState currNodeState = currNodes.get(endpoint);
 
                         //update the old node state by the new one
-                        if (currNodeState.isConnected() || !prevNodeState.isConnected()) {
+
+                        if (!prevNodeState.isConnected() && currNodeState.isConnected()) {
                             actualState.put(endpoint, currNodeState);
-                        } else {
+                        } else if (prevNodeState.isConnected() && currNodeState.isConnected()) {
+                            Map<String, NodeConnectivity.ConnectionStatus> mergedStates = new HashMap<>(
+                                    prevNodeState
+                                            .getConnectivity()
+                                            .getConnectivity());
+
+                            currNodeState.getConnectivity().getConnectivity().forEach((k, v) ->
+                                    mergedStates.merge(k, v, (item1, item2) ->
+                                            item1 == NodeConnectivity.ConnectionStatus.OK
+                                                    || item2 == NodeConnectivity.ConnectionStatus.OK ?
+                                                    NodeConnectivity.ConnectionStatus.OK :
+                                                    NodeConnectivity.ConnectionStatus.FAILED));
+
+                            NodeConnectivity nodeConnectivity = NodeConnectivity.builder()
+                                    .endpoint(endpoint)
+                                    .type(currNodeState.getConnectivity().getType())
+                                    .connectivity(ImmutableMap.copyOf(mergedStates))
+                                    .epoch(Math.max(prevNodeState.getConnectivity().getEpoch(), currNodeState.getConnectivity().getEpoch()))
```

Comment:
Can we add some tests that exercises different epochs between node states? When the epochs differ, it isn't clear that we can assume that `prevNodes` has information about an endpoint just because `currNodes` does. Particularly, when interleaved with add-node/remove-node workflows.

Reply:
The existing design/code is not meant to aggregate heart beating information across epochs. So instead of `Math.max`, we should probably assert the invariant or log a warning. If you look at `pollAsync`, it collects responses from `sendNodeStateRequest` which is stamped with the same epoch. The implication of this is all collected responses that return a `NodeState` will have the same epoch.